packages/program/src/functions/base-invocation-scope.ts

```diff
@@ -226,13 +232,58 @@ export class BaseInvocationScope<TReturn = any> {
    *
    * @returns The current instance of the class.
    */
-  async fundWithRequiredCoins(fee: BN) {
+  async fundWithRequiredCoins() {
+    const transactionRequest = await this.getTransactionRequest();
+
+    const { maxFee, gasUsed, minGasPrice, estimatedInputs, estimatedOutputs } =
+      await this.getTransactionCost();
+
+    this.setDefaultTxParams(transactionRequest, minGasPrice, gasUsed);
+
+    transactionRequest.outputs = estimatedOutputs;
+
     // Clean coin inputs before add new coins to the request
-    this.transactionRequest.inputs = this.transactionRequest.inputs.filter(
-      (i) => i.type !== InputType.Coin
+    this.transactionRequest.inputs = estimatedInputs.filter((i) => i.type !== InputType.Coin);
+
+    await this.program.account?.fund(this.transactionRequest, this.requiredCoins, maxFee);
+
+    // update predicate inputs with estimated predicate-related info because the funding removes it
+    this.transactionRequest.inputs.forEach((input) => {
+      let estimatedInput: TransactionRequestInput | undefined;
+      switch (input.type) {
+        case InputType.Contract:
+          return;
+        case InputType.Coin:
+          estimatedInput = estimatedInputs.find(
+            (x) => x.type === InputType.Coin && x.owner === input.owner
+          );
+          break;
+        case InputType.Message:
+          estimatedInput = estimatedInputs.find(
+            (x) => x.type === InputType.Message && x.sender === input.sender
+          );
+          break;
+        default:
+          break;
+      }
+      if (estimatedInput && 'predicate' in estimatedInput) {
+        input.predicate = estimatedInput.predicate;
+        input.predicateData = estimatedInput.predicateData;
+        input.predicateGasUsed = estimatedInput.predicateGasUsed;
```

Comment:
Because we are adding inputs in the following way: https://github.com/FuelLabs/fuels-ts/blob/ca859efc525dacb1d93c11ff535e438acaec6439/packages/account/src/providers/transaction-request/transaction-request.ts#L342-L343
`'predicate' in estimatedInput` will always be true since when the input it's not a predicate the value will be:
```ts
{
  ...
  predicate: undefined,
  predicateData: undefined,
}
```
 and this means that the code will enter inside this loop unnecessary. 

We can change it to something like:
```ts
if (
        estimatedInput &&
        'predicateGasUsed' in estimatedInput &&
        bn(estimatedInput.predicateGasUsed).gt(0)
      )
```
Or to any other logic that will do the trick.

Reply:
https://github.com/FuelLabs/fuels-ts/pull/1767/commits/1844dac8de87d5d701115d4d5a2d4f1460db168f